File: src/main/java/org/example/tree/SameTree.java
package org.example.tree;

public class SameTree {

    public boolean isSameTree(TreeNode p, TreeNode q) {
        if (p == null && q == null) return true;

        if (p == null) return false;
        if (q == null) return false;
        boolean leftSub = p.val == q.val && isSameTree(p.left, q.left);
        boolean rightSub = p.val == q.val && isSameTree(p.right, q.right);

        return leftSub & rightSub;
    }


}

File: src/main/java/org/example/tree/SameTree.py
from typing import Optional

from org.example.tree.TreeNode import TreeNode


class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:

        def dfs(p, q):
            if not p and not q: return True
            if (p and not q) or (not p and q): return False

            left = (p.val == q.val and dfs(p.left, q.left))
            right = (p.val == q.val and dfs(p.right, q.right))

            return left and right
        return dfs(p, q)